File: src/main/java/chiloven/lukosbot2/commands/music/MusicCommand.java
package chiloven.lukosbot2.commands.music;

import chiloven.lukosbot2.commands.BotCommand;
import chiloven.lukosbot2.commands.music.provider.MusicProvider;
import chiloven.lukosbot2.commands.music.provider.SoundCloudMusicProvider;
import chiloven.lukosbot2.commands.music.provider.SpotifyMusicProvider;
import chiloven.lukosbot2.config.AppProperties;
import chiloven.lukosbot2.core.CommandSource;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;

import java.util.List;
import java.util.function.Predicate;

/**
 * The music command for querying music info from streaming platforms.
 *
 * @author Chiloven945
 */
public class MusicCommand implements BotCommand {
    private final MusicProvider spotify;
    private final MusicProvider soundCloud;

    public MusicCommand(AppProperties.Music music) {
        AppProperties.Music.Spotify sp = music.getSpotify();
        this.spotify = sp.isEnabled() && sp.getClientId() != null && !sp.getClientId().isBlank() && sp.getClientSecret() != null && !sp.getClientSecret().isBlank()
                ? new SpotifyMusicProvider(sp.getClientId(), sp.getClientSecret())
                : null;

        AppProperties.Music.SoundCloud sc = music.getSoundcloud();
        this.soundCloud = (sc.isEnabled() && sc.getClientId() != null && !sc.getClientId().isBlank())
                ? new SoundCloudMusicProvider(sc.getClientId())
                : null;
    }

    @Override
    public String name() {
        return "music";
    }

    @Override
    public String description() {
        return "从流媒体平台查询音乐信息";
    }

    @Override
    public String usage() {
        return """
                用法：
                /music <query>
                /music <platform> <query>
                /music link <link>
                示例：
                /music Never Gonna Give You Up
                /music spotify
                /music link
                """;
    }

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
                LiteralArgumentBuilder.<CommandSource>literal(name())
                        // /music
                        .executes(ctx -> {
                            ctx.getSource().reply(usage());
                            return 1;
                        })
                        // /music link <link>
                        .then(LiteralArgumentBuilder.<CommandSource>literal("link")
                                .then(RequiredArgumentBuilder
                                        .<CommandSource, String>argument("link", StringArgumentType.greedyString())
                                        .executes(ctx -> runLink(
                                                ctx.getSource(),
                                                StringArgumentType.getString(ctx, "link")
                                        ))
                                )
                        )
                        // /music <query>
                        // /music <platform> <query>
                        .then(RequiredArgumentBuilder
                                .<CommandSource, String>argument("first", StringArgumentType.word())
                                // Only one argument: treat as query
                                .executes(ctx -> runSearch(
                                        ctx.getSource(),
                                        StringArgumentType.getString(ctx, "first"),
                                        null
                                ))
                                // The second argument eats the rest: could be `<platform> <query>` or just a multi-word query
                                .then(RequiredArgumentBuilder
                                        .<CommandSource, String>argument("rest", StringArgumentType.greedyString())
                                        .executes(ctx -> {
                                            CommandSource src = ctx.getSource();
                                            String first = StringArgumentType.getString(ctx, "first");
                                            String rest = StringArgumentType.getString(ctx, "rest");

                                            Predicate<String> isPlatformName = s -> {
                                                if (s == null) return false;
                                                String p = s.toLowerCase();
                                                return List.of("spotify", "soundcloud", "sc").contains(p);
                                            };

                                            return isPlatformName.test(first)
                                                    // /music <platform> <query>
                                                    ? runSearch(src, rest, first)
                                                    // /music <query-with-space>
                                                    : runSearch(src, first + " " + rest, null);
                                        })
                                )
                        )
        );
    }

    private int runSearch(CommandSource src, String query, String platformName) {
        try {
            MusicProvider provider = pickProvider(platformName);
            if (provider == null) {
                src.reply("未配置可用的音乐平台（Spotify / SoundCloud），请在 config/application.yml 中设置 lukos.music.*");
                return 0;
            }
            TrackInfo info = provider.searchTrack(query);
            if (info == null) {
                src.reply("没有找到匹配的歌曲。");
                return 0;
            }
            src.reply(info.formatted());
            return 1;
        } catch (Exception e) {
            src.reply("查询失败：" + e.getMessage());
            return 0;
        }
    }

    private int runLink(CommandSource src, String link) {
        try {
            MusicProvider provider = detectProviderByLink(link);
            if (provider == null) {
                src.reply("无法识别链接所属平台，仅支持 Spotify / SoundCloud。");
                return 0;
            }
            TrackInfo info = provider.resolveLink(link);
            if (info == null) {
                src.reply("无法从链接解析到歌曲信息。");
                return 0;
            }
            src.reply(info.formatted());
            return 1;
        } catch (Exception e) {
            src.reply("解析链接失败：" + e.getMessage());
            return 0;
        }
    }

    private MusicProvider pickProvider(String platformName) {
        if (platformName == null) {
            // Spotify default, then SoundCloud
            if (spotify != null) return spotify;
            return soundCloud;
        }
        String p = platformName.toLowerCase();
        return switch (p) {
            case "s", "spotify" -> spotify;
            case "sc", "soundcloud" -> soundCloud;
            default -> null;
        };
    }

    private MusicProvider detectProviderByLink(String link) {
        if (link == null) return null;
        String s = link.toLowerCase();
        if (s.contains("open.spotify.com")) return spotify;
        if (s.contains("soundcloud.com")) return soundCloud;
        return null;
    }
}

File: src/main/java/top/chiloven/lukosbot2/commands/bot/music/MusicCommand.kt
package top.chiloven.lukosbot2.commands.bot.music

import org.apache.logging.log4j.LogManager
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Service
import top.chiloven.lukosbot2.commands.IBotCommand
import top.chiloven.lukosbot2.commands.bot.music.provider.IMusicProvider
import top.chiloven.lukosbot2.commands.bot.music.provider.SoundCloudMusicProvider
import top.chiloven.lukosbot2.commands.bot.music.provider.SpotifyMusicProvider
import top.chiloven.lukosbot2.config.CommandConfigProp
import top.chiloven.lukosbot2.core.command.bot.CommandSource
import top.chiloven.lukosbot2.core.command.definition.dsl.botCommand

@Service
@ConditionalOnProperty(
    prefix = "lukos.commands.control",
    name = ["music"],
    havingValue = "true",
    matchIfMissing = true
)
class MusicCommand(ccp: CommandConfigProp) : IBotCommand {

    private val log = LogManager.getLogger(MusicCommand::class.java)

    private val spotify: IMusicProvider? = ccp.music.spotify.let { sp ->
        if (sp.enabled && sp.clientId.isNotBlank() && sp.clientSecret.isNotBlank()) SpotifyMusicProvider(
            sp.clientId,
            sp.clientSecret
        ) else null
    }
    private val soundCloud: IMusicProvider? = ccp.music.soundcloud.let { sc ->
        if (sc.enabled && sc.clientId.isNotBlank()) SoundCloudMusicProvider(sc.clientId) else null
    }

    private val commandDefinition = botCommand("music") {
        description = "从流媒体平台查询音乐信息"

        literal("link") {
            description = "通过链接解析歌曲"
            raw("link") { link -> runLink(source, link) }
            param("link", "Spotify 或 SoundCloud 链接")
        }

        raw("query", required = false) { query ->
            if (query.isBlank()) {
                sendUsage(source)
                return@raw
            }

            val tokens = query.split(" ", limit = 2)
            if (isPlatformToken(tokens[0])) {
                runSearch(source, tokens.getOrElse(1) { "" }, tokens[0])
            } else {
                runSearch(source, query, null)
            }
        }

        param("query", "搜索关键字（支持空格）")

        example(
            "music Never Gonna Give You Up",
            "music spotify Never Gonna Give You Up",
            "music link https://open.spotify.com/track/xxxxxxxx"
        )
    }

    override fun definition() = commandDefinition

    private fun runSearch(src: CommandSource, query: String, platformToken: String?): Int {
        return try {
            val provider = pickProvider(platformToken) ?: run {
                src.reply("音乐平台暂未配置或不可用。")
                return 0
            }
            val info = provider.searchTrack(query) ?: run {
                src.reply("没有找到匹配的歌曲。")
                return 0
            }

            src.reply(info.formatted())
            1
        } catch (e: Exception) {
            src.reply("查询失败：${e.message ?: "未知错误"}")
            log.warn("Music search failed: query='{}', platform='{}'", query, platformToken, e)
            0
        }
    }

    private fun runLink(src: CommandSource, link: String): Int {
        return try {
            val provider = detectProviderByLink(link) ?: run {
                src.reply("无法识别链接所属平台。")
                return 0
            }
            val info = provider.resolveLink(link) ?: run {
                src.reply("无法从链接解析到歌曲信息。")
                return 0
            }

            src.reply(info.formatted())
            1
        } catch (e: Exception) {
            src.reply("解析链接失败：${e.message ?: "未知错误"}")
            log.warn("Music link failed: link='{}'", link, e)
            0
        }
    }

    private fun pickProvider(platformToken: String?) = when (platformToken?.lowercase()) {
        null, "" -> spotify ?: soundCloud
        "s", "spotify" -> spotify
        "sc", "soundcloud" -> soundCloud
        else -> null
    }

    private fun detectProviderByLink(link: String?) = when {
        link == null -> null
        "open.spotify.com" in link.lowercase() -> spotify
        "soundcloud.com" in link.lowercase() -> soundCloud
        else -> null
    }

    private fun isPlatformToken(s: String?) =
        s?.lowercase() in setOf("spotify", "soundcloud", "sc", "s")

}
